substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -751,157 +749,279 @@ pub mod pallet {
+						let free =
+							T::Assets::reducible_balance(asset.clone(), who, preservation, Polite);
+						ensure!(free >= value, TokenError::NotExpendable);
+					}
+					T::Assets::withdraw(asset, who, value, Exact, preservation, Polite)
+						.map(|c| c.into())
+				},
+				MultiAssetIdConversionResult::Native => {
+					if preservation == Preserve {
+						// TODO drop the ensure! when this issue addressed
+						// https://github.com/paritytech/polkadot-sdk/issues/1698
+						let free = T::Currency::reducible_balance(who, preservation, Polite);
+						ensure!(free >= value, TokenError::NotExpendable);
+					}
+					T::Currency::withdraw(who, value, Exact, preservation, Polite).map(|c| c.into())
+				},
+				MultiAssetIdConversionResult::Unsupported(_) =>
+					Err(Error::<T>::UnsupportedAsset.into()),
+			}
 		}
 
-		/// Swap assets along a `path`, depositing in `send_to`.
-		pub(crate) fn do_swap(
-			sender: T::AccountId,
-			amounts: &Vec<T::AssetBalance>,
-			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
-			send_to: T::AccountId,
+		/// Swap assets along the `path`, withdrawing from `sender` and depositing in `send_to`.
+		///
+		/// Note: It's assumed that the provided `path` is valid.
+		///
+		/// WARNING: This may return an error after a partial storage mutation. It should be used
+		/// only inside a transactional storage context and an Err result must imply a storage
+		/// rollback.
+		fn swap(
+			sender: &T::AccountId,
+			path: &BalancePath<T>,
+			send_to: &T::AccountId,
 			keep_alive: bool,
 		) -> Result<(), DispatchError> {
-			ensure!(amounts.len() > 1, Error::<T>::CorrespondenceError);
-			if let Some([asset1, asset2]) = &path.get(0..2) {
-				let pool_id = Self::get_pool_id(asset1.clone(), asset2.clone());
-				let pool_account = Self::get_pool_account(&pool_id);
-				// amounts should always contain a corresponding element to path.
-				let first_amount = amounts.first().ok_or(Error::<T>::CorrespondenceError)?;
-
-				Self::transfer(asset1, &sender, &pool_account, *first_amount, keep_alive)?;
-
-				let mut i = 0;
-				let path_len = path.len() as u32;
-				for assets_pair in path.windows(2) {
-					if let [asset1, asset2] = assets_pair {
-						let pool_id = Self::get_pool_id(asset1.clone(), asset2.clone());
-						let pool_account = Self::get_pool_account(&pool_id);
-
-						let amount_out =
-							amounts.get((i + 1) as usize).ok_or(Error::<T>::CorrespondenceError)?;
-
-						let to = if i < path_len - 2 {
-							let asset3 = path.get((i + 2) as usize).ok_or(Error::<T>::PathError)?;
-							Self::get_pool_account(&Self::get_pool_id(
+			let (asset_in, amount_in) = path.first().ok_or(Error::<T>::InvalidPath)?;
+			let credit_in = Self::withdraw(asset_in, sender, *amount_in, keep_alive)?;
+
+			let credit_out = Self::credit_swap(credit_in, path).map_err(|(_, e)| e)?;
+			Self::resolve(send_to, credit_out).map_err(|_| Error::<T>::BelowMinimum)?;
```

Comment:
the `Self::resolve()` method's name doesn't describe/suggest what the function does under the hood, it only complicates the reading

Reply:
it just mirrors the fungbiles resolve function. this abstraction will be dropped when the bug requiring it will be addressed.